crates/fuel-core/src/service/genesis/workers.rs

```diff
@@ -0,0 +1,311 @@
+        finished_signal: Option<Arc<Notify>>,
+    ) -> GenesisRunner<Handler<T>, I, Database>
+    where
+        Handler<T>: ProcessState<Item = T>,
+        I: IntoIterator<Item = anyhow::Result<Group<T>>>,
+    {
+        let handler = Handler::new(self.block_height);
+        let database = self.db.clone();
+        GenesisRunner::new(
+            finished_signal,
+            self.cancel_token.clone(),
+            handler,
+            data,
+            database,
+        )
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct Handler<T> {
+    output_index: u64,
+    block_height: BlockHeight,
+    phaton_data: PhantomData<T>,
+}
+
+impl<T> Handler<T> {
+    pub fn new(block_height: BlockHeight) -> Self {
+        Self {
+            output_index: 0,
+            block_height,
+            phaton_data: PhantomData,
+        }
+    }
+}
+
+impl ProcessState for Handler<CoinConfig> {
+    type Item = CoinConfig;
+
+    fn process(
+        &mut self,
+        group: Vec<Self::Item>,
+        tx: &mut Database,
+    ) -> anyhow::Result<()> {
+        group
+            .into_iter()
+            .try_for_each(|coin| {
+                let root = init_coin(tx, &coin, self.output_index, self.block_height)?;
+        tx.add_coin_root(root)?;
+
+        self.output_index = self.output_index
+                .checked_add(1)
+                .expect("The maximum number of UTXOs supported in the genesis configuration has been exceeded.");
+
+        Ok(())
+            })
+    }
+
+    fn genesis_resource() -> GenesisResource {
+        GenesisResource::Coins
+    }
+}
+
+impl ProcessState for Handler<MessageConfig> {
+    type Item = MessageConfig;
+
+    fn process(
+        &mut self,
+        group: Vec<Self::Item>,
+        tx: &mut Database,
+    ) -> anyhow::Result<()> {
+        group.into_iter().try_for_each(|message| {
+            let root = init_da_message(tx, &message)?;
+            tx.add_message_root(root)?;
+            Ok(())
+        })
+    }
+
+    fn genesis_resource() -> GenesisResource {
+        GenesisResource::Messages
+    }
+}
+
+impl ProcessState for Handler<ContractConfig> {
+    type Item = ContractConfig;
+
+    fn process(
+        &mut self,
+        group: Vec<Self::Item>,
+        tx: &mut Database,
+    ) -> anyhow::Result<()> {
+        group
+            .into_iter()
+            .try_for_each(|contract| {
+                init_contract(tx, &contract, self.output_index, self.block_height)?;
+        tx.add_contract_id(contract.contract_id)?;
+
+        self.output_index = self.output_index
+                .checked_add(1)
+                .expect("The maximum number of UTXOs supported in the genesis configuration has been exceeded.");
+
+        Ok::<(), anyhow::Error>(())
+            })
+    }
+
+    fn genesis_resource() -> GenesisResource {
+        GenesisResource::Contracts
+    }
+}
+
+// wrapper for ContractId to be used in the ProcessState trait
+
+impl ProcessState for Handler<ContractId> {
```

Comment:
Hmm, it seems part of the comments were missed after I pushed changes=(

We discussed on the call that it would be nice to use something else than `ContractId`. Maybe `ContractRootCalculatorConfig` or something with a similar idea to be along with another implementation of this trait for the config related only to the regenesis.

Reply:
I've changed the implementation to use the already existing `ContractsInfo` table since it already contains the loaded contract ids. 
IIUC the problem with `ContractId` is that the the handler implementation for it comes as a surprise but it differs from the other *Config items in that it isn't read from a file.
I've added the `GenesisImportedContractId` alias to make it's purpose more obvious. What do you think?